### ISSUE TITLE
Mockery\Exception\BadMethodCallException: Received Mockery_0_*::__unset(), but no expectations were specified

### ISSUE THREAD

**Issue by josefsabl** [padraic/mockery#926]
I am new to mockery and I am testing very trivial abstract class by mocking it. Let's say my class looks like this:

```php
abstract class Store {
    /** @var array */
    private $data;

    //....

    /** @param string $key */
    public function __unset(string $key) {
        unset($this->data[$key]);
    }
}
```

As it is abstract, I mock it like this:

```php
    public function setUp() {
        $this->document = \Mockery::mock(Store::class);
    }
```

And then test it like this:

```php
    public function testUnset() {
            unset($this->document->foo);
            Assert::false(isset($this->document->foo));
    }
```

But I get this weird exception.

```
Mockery\Exception\BadMethodCallException: Received Mockery_0__Store::__unset(), but no expectations were specified
```

I wasn't able to trace the problem with Xdebug, as it can't go into the mocked object obviously.

What am I doing wrong? Or is it a bug? The `__get`, `__set` and `__isset` magic methods work just fine as well as `__construct` does.

**Comment by davedevelopment** [comment]
Honestly, I'm not sure :flushed: 

It's not something I've ever delved in to. Historically, I know we tried to support `__get` and `__set`, but something about them was never quite right, so they dropped the support. This was before I was deeply involved in the project.

If someone has time to take a closer look it would be appreciated.

Regarding your debugging with xdebug, if you put this somewhere in your test bootstrap:
``` php

// tests/bootstrap.php

Mockery::setLoader(new Mockery\Loader\RequireLoader(sys_get_temp_dir()));

```
That will tell mockery to dump it's generated mock code in a file somewhere in `/tmp` and then `include` it, rather than `eval`. That way, your editor and debugger might be able to take a look at what is happening.

**Comment by josefsabl** [comment]
After refactoring my code the class is not abstract anymore so I don't use mockery to test it. But I will try to replicate the problem and use your suggestion with loader to find out more.

**Comment by robertbasic** [comment]
@josefsabl maybe try and make the mock a [partial mock](http://docs.mockery.io/en/latest/reference/partial_mocks.html)? So that way the original `__unset` method gets called.

Also, why do you think the error message is weird? The `Mockery\Exception\BadMethodCallException: Received Mockery_0__Store::__unset(), but no expectations were specified` error message says that it tried to call the `__unset` method, but it was not defined on the mock object. Is there a way we could improve it?

**Comment by josefsabl** [comment]
> Is there a way we could improve it?

Please understand that I am a complete noob to Mockery and didn't RTFM. AT ALL :-) And I am NOT complaining here at all.

But I think I understand the question and this is about UX, right? I will try to give some, hopefully useful, feedback.

When I called the `\Mockery::mock(Store::class);` I expected to get the concrete class instance extending the abstract and thought that I will only need to mock the method abstracts. Actually, I didn't think about it much as I didn't need to test the abstracts.

And why did I find this exception message is weird? I have absolutely no idea, what the `expectation` is :-) If I understand it correctly. And I assume I am not. I think I would be more happy with something like: "__unset method does not exist in the mock. Did you forget to mock it? Maybe you wanted to use partial mock instead".

To sum it: Exception messages are part of user interface in case of libraries and should use adequate language.

**Comment by robertbasic** [comment]
Yes, it is about UX/DX (developer experience). I've been using Mockery for more than 5 years so I understand what it means when I get an exception like that, but I also understand that it can be confusing to someone new to Mockery. Thanks for the suggestion, I'll see to improving it :)
